lib/utils/set/set.go

```diff
@@ -0,0 +1,128 @@
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package set
+
+import "maps"
+
+// Set models a collection of unique elements. Its implemented as the classic
+// Go `map[T]struct{}` pattern and has the same underlying representation. This
+// has the nice property of preserving the map-like reference semantics. You can
+// even iterate over the set with `range`.
+type Set[T comparable] map[T]struct{}
+
+// New constructs a set from an arbitrary collection of elements
+func New[T comparable](elements ...T) Set[T] {
+	s := NewWithCapacity[T](len(elements))
+	s.Add(elements...)
+	return s
+}
+
+// NewWithCapacity constructs a new, empty set with an given capacity hint.
+func NewWithCapacity[T comparable](n int) Set[T] {
+	return make(map[T]struct{}, n)
+}
+
+// Add expands the set to include the supplied elements (if not already included),
+// returning a reference to the mutated set.
+func (s Set[T]) Add(elements ...T) Set[T] {
+	for _, element := range elements {
+		s[element] = struct{}{}
+	}
+	return s
+}
+
+// Len returns the number of elements in the set.
+func (s Set[T]) Len() int {
+	return len(s)
+}
+
+// Union updates the set to be the union of the original set and `other`
+func (s Set[T]) Union(others ...Set[T]) {
+	for _, other := range others {
+		for element := range other {
+			s[element] = struct{}{}
+		}
+	}
+}
+
+// Remove deletes elements from the set, returning a reference to the mutated
+// set. Attempting to remove a non-existent element from the set is not
+// considered an error.
+func (s Set[T]) Remove(elements ...T) Set[T] {
+	for _, element := range elements {
+		delete(s, element)
+	}
+	return s
+}
+
+// Contains implements a membership test for the Set.
+func (s Set[T]) Contains(element T) bool {
+	_, present := s[element]
+	return present
+}
+
+// Clone creates a new, independent copy of the Set.
+func (s Set[T]) Clone() Set[T] {
+	return maps.Clone(s)
+}
+
+// Subtract removes all elements in `other` from the set (i.e `s` becomes the
+// Set Difference of `s` and `other`), returning a reference to the mutated set.
+func (s Set[T]) Subtract(other Set[T]) Set[T] {
+	for k := range other {
+		delete(s, k)
+	}
+	return s
+}
+
+// Intersection updates the set to contain the similarity between the set and `other`
+func (s Set[T]) Intersection(other Set[T]) {
+	for b := range s {
+		if !other.Contains(b) {
+			s.Remove(b)
+		}
+	}
+}
+
+// Elements returns the elements in the set. Order of the elements is undefined.
+//
+// NOTE: Due to the underlying map type, a set can be naturally ranged over like
+// a map, for example:
+//
+//	alphabet := NewSet("alpha", "beta", "gamma")
+//	for l := range alphabet {
+//		fmt.Printf("%s is a letter\n", l)
+//	}
+//
+// Prefer using the natural range iteration where possible
+func (s Set[T]) Elements() []T {
+	elements := make([]T, 0, len(s))
```

Comment:
May be better for a different PR, but could modernize.

```suggestion
	return slices.Collect(maps.Keys(elements))
```